### DISCUSSION TITLE
How to stop execution when using @runno/wasi directly?

### DISCUSSION THREAD

**Discussion by fnando** [taybenlor/runno#288] [Q&A] [answered]
Hello! Thanks for this nice project. 

I'm using the @runno/wasi directly without the built-in ui and was thinking how would it be possible to stop the execution. I tried a few things but couldn't really figure out how to do it.

Any hints? This is how I'm currently running the code:

```typescript
const code = "puts RUBY_VERSION"

const result = await WASI.start(fetch("/wasm/ruby-3.3.0.wasm"), {
  args: ["ruby", "-r", "/ruby/setup.rb", "/code/ruby.rb"],
  env: env || {},
  stdout: (content) => console.log(content),
  stderr: (content) => console.log(content),
  stdin: () => null,
  fs: {
    "/code/ruby.rb": {
      timestamps,
      path: "/code/ruby.rb",
      mode: "string",
      content: code,
    },

    "/ruby/setup.rb": {
      timestamps,
      path: "/ruby/setup.rb",
      mode: "string",
      content: `
$stdout.sync = true
$stderr.sync = true
`,
    },
  },
});
```

Thanks!

**Comment by fnando** [comment, marked as answer]
Well, I made it work with `WASIWorkerHost`. If the execution doesn't complete in a couple of seconds, I kill the host with `host.kill()`.

Is that the best way of doing it? Any side effects on this approach?

**Reply by taybenlor** [reply]
That's going to be the best way to do it unfortunately! When running WebAssembly synchronously we are relinquishing control of the thread over to it. I would like to create some kind of sandboxing controls to help with this, but right now I can't think of a neat way to do it without modifying the WebAssembly binary.
